File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        uniques = set()
        n = len(s)
        left = 0
        max_len = 0

        for right in range(n):
            while s[right] in uniques:
                uniques.remove(s[left])
                left += 1
                
            uniques.add(s[right])
            max_len = max(max_len, right-left+1)
        
        return max_len
            
                
                

                
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func max(a, b int) int {
    if a > b {
        return a;
    } else {
        return b;
    }
}

func lengthOfLongestSubstring(s string) int {
    var left int = 0;
    var counter = map[byte]int {}; // a map from value to frequency
    var maxSubstringLength int = 0;

    for right := 0; right < len(s); right++ {
        // increment counter
        counter[s[right]] ++;

        for counter[s[right]] > 1 {
            counter[s[left]] --;
            left ++;
        }
        maxSubstringLength = max(maxSubstringLength, right - left + 1);
    }
    return maxSubstringLength;
}
